lib/src/tasks/coverage/api.dart

```diff
@@ -406,45 +417,101 @@ class CoverageTask extends Task {
     String _testsPassedPattern = 'All tests passed!';
 
     if (isBrowserTest) {
-      // Run the test in content-shell.
-      String executable = 'content_shell';
-      List args = [htmlFile.path];
-      _coverageOutput.add('');
-      _coverageOutput.add('Running test suite ${file.path}');
-      _coverageOutput.add('$executable ${args.join(' ')}\n');
-      TaskProcess process =
-          _lastTestProcess = new TaskProcess('content_shell', args);
+      PubServeTask pubServeTask;
 
-      // Content-shell dumps render tree to stderr, which is where the test
-      // results will be. The observatory port should be output to stderr as
-      // well, but it is sometimes malformed. In those cases, the correct
-      // observatory port is output to stdout. So we listen to both.
-      int observatoryPort;
-      process.stdout.listen((line) {
-        _coverageOutput.add('    $line');
-        if (line.contains(_observatoryPortPattern)) {
-          Match m = _observatoryPortPattern.firstMatch(line);
-          observatoryPort = int.parse(m.group(2));
-        }
-      });
-      await for (String line in process.stderr) {
-        _coverageOutput.add('    $line');
-        if (line.contains(_observatoryFailPattern)) {
-          throw new CoverageTestSuiteException(file.path);
-        }
-        if (line.contains(_observatoryPortPattern)) {
-          Match m = _observatoryPortPattern.firstMatch(line);
-          observatoryPort = int.parse(m.group(2));
-        }
-        if (line.contains(_testsFailedPattern)) {
-          throw new CoverageTestSuiteException(file.path);
+      try {
+        String testPath;
+        if (pubServe) {
+          _coverageOutput.add('Starting pub server...');
+
+          // Start `pub serve` on the `test` directory.
+          pubServeTask = startPubServe(additionalArgs: ['test']);
+
+          _coverageOutput.add('::: ${pubServeTask.command}');
+          String indentLine(String line) => '    $line';
+
+          var startupLogFinished = new Completer();
+          pubServeTask.stdOut
+              .transform(until(startupLogFinished.future))
+              .map(indentLine)
+              .listen(_coverageOutput.add);
+          pubServeTask.stdErr
+              .transform(until(startupLogFinished.future))
+              .map(indentLine)
+              .listen(_coverageErrorOutput.add);
+
+          PubServeInfo serveInfo =
+              await pubServeTask.serveInfos.firstWhere((info) {
+            return path.isWithin(info.directory, htmlFile.path);
+          }).timeout(const Duration(seconds: 10), onTimeout: () {
+            throw new TimeoutException(
+                'failed to detect `pub serve` directory containing ${htmlFile.path}');
+          });
```

Comment:
Hmm, now that I think about it, if we're passing in `additionalArgs: ['test']` to `startPubServe`, this stream will almost assuredly contain only one event.

This should probably be changed to:
```dart
PubServeInfo serveInfo = await pubServeTask.serveInfos.first;
if (!path.isWithin(serveInfo.directory, htmlFile.path)) {
  throw '`pub serve` directory `${serveInfo.directory}` does not contain test file `${htmlFile.path}`.';
}
```